lib/src/scanner/mod.rs

```diff
@@ -346,6 +374,25 @@ impl<'r> Scanner<'r> {
             ScannedData::Mmap(mapped_file)
         };
 
+        Ok(data)
+    }
+
+    /// Scans a file.
+    pub fn scan_file<'a, 'target, 'meta>(
+        &'a mut self,
+        scan_input: &ScanInput<'target, 'meta>,
```

Comment:
This is a backward incompatible change in the API that is not required nor desirable. In the vast majority of cases the API user just wants to scan a file and don't want to be bothered with additional details like the metadata. Moreover, if we would pass the metadata to this function, why not simply add a new argument? Like this:

```rust
pub fn scan_file<P: AsRef<Path>, M: AsRef<Path>>(&self, path: P, meta: Option<M>)
```

The introduction of  the`ScanInput` struct to the API seems unnecessary. However, this is not the most problematic issue with this design. The main problem is that the additional metadata is not always the same for every module. The `-x` argument in `yara` has the following form: `--module-data=MODULE=FILE`, where `MODULE` is a the module name and `FILE` the path to the file passed to that module. This option can be used multiple times, passing different files to each module.  You can have modules `foo` and `bar`, each expecting a different kind of metadata and invoke YARA like:

```shell
yara -module-data=foo=file1.json --module-data=bar=file2.json ...
```

The current design doesn't cover this case, all modules will see exactly the same metadata.

My proposal is adding a `set_module_meta` to the `Scanner` object, which works in a similar way to [set_module_output](https://docs.rs/yara-x/latest/yara_x/struct.Scanner.html#method.set_module_output).

Before calling `scan`, the user calls `set_module_meta` in order to specify the metadata that will be passed to a specific module. This function will receive both the module's name and a slice with the metadata for that module. Once `scan` is called, the metadata for that module is cleared and in subsequent calls to `scan` the module won't receive any metadata unless the user calls `set_module_input` again before calling `scan`.

Of course, this requires changing the signature for the module's main function, but instead of introducing a new struct `ScanInputRaw` it could simply be changed to: 

```
main(data: &[u8], meta: Option<&[u8]>)
```

Reply:
thx for the feedback

implemented in [b9a6140](https://github.com/VirusTotal/yara-x/pull/183/commits/b9a61401036315f758fb1ba4138e5085b68de42f)